File: sdk/runanywhere-swift/Sources/RunAnywhere/Foundation/DeviceIdentity/PersistentDeviceIdentity.swift
import Foundation
#if os(iOS) || os(tvOS)
import UIKit
#endif

/// Comprehensive device identity manager that provides persistent UUIDs across app reinstalls
/// Uses multiple fallback mechanisms to ensure device identification persistence
public class PersistentDeviceIdentity {

    private static let logger = SDKLogger(category: "PersistentDeviceIdentity")
    private static let keychainManager = KeychainManager.shared

    /// Get a persistent device UUID that survives app reinstalls
    /// Uses a multi-layered approach for maximum persistence
    public static func getPersistentDeviceUUID() -> String {
        logger.debug("Attempting to retrieve persistent device UUID")

        // Strategy 1: Try to get from persistent keychain (survives app reinstalls)
        if let persistentUUID = keychainManager.retrieveDeviceUUID() {
            logger.info("Retrieved device UUID from persistent keychain")
            return persistentUUID
        }

        // Strategy 2: Use Apple's identifierForVendor (stable but can change)
        if let vendorUUID = getVendorBasedUUID() {
            logger.info("Retrieved device UUID from vendor identifier")
            // Store this in persistent keychain for future use
            try? keychainManager.storeDeviceUUID(vendorUUID)
            return vendorUUID
        }

        // Strategy 3: Generate new UUID and store persistently
        let newUUID = generateAndStoreNewUUID()
        logger.info("Generated new device UUID")
        return newUUID
    }

    /// Get device fingerprint for additional validation
    /// This provides device characteristics that can help verify identity
    public static func getDeviceFingerprint() -> String {
        var components: [String] = []

        // Hardware characteristics (these don't change)
        let processInfo = ProcessInfo.processInfo
        components.append("mem:\(processInfo.physicalMemory)")

        #if arch(arm64)
        components.append("arch:arm64")
        #elseif arch(x86_64)
        components.append("arch:x86_64")
        #else
        components.append("arch:unknown")
        #endif

        // Use DeviceKitAdapter if available for chip info
        let adapter = DeviceKitAdapter()
        let processorInfo = adapter.getProcessorInfo()
        components.append("chip:\(processorInfo.chipName)")
        components.append("cores:\(processorInfo.coreCount)")

        // OS info (major version only, as minor versions change)
        let osVersion = processInfo.operatingSystemVersion
        components.append("os:\(osVersion.majorVersion)")

        // Create fingerprint hash
        let fingerprintString = components.joined(separator:"|")
        let fingerprint = sha256(fingerprintString)

        // Store fingerprint for future validation
        try? keychainManager.storeDeviceFingerprint(fingerprint)

        return fingerprint
    }

    /// Validate if a device UUID still represents the same physical device
    public static func validateDeviceUUID(_ uuid: String) -> Bool {
        // For now, we trust any UUID that's properly formatted
        // In the future, this could compare against device fingerprints
        return uuid.count == 36 && uuid.contains("-")
    }

    // MARK: - Private Methods

    // Note: The keychain storage/retrieval is now handled by KeychainManager
    // KeychainManager uses kSecAttrAccessibleWhenUnlockedThisDeviceOnly for security
    // which provides good persistence across app launches

    /// Get vendor-based UUID (Apple's recommended approach)
    private static func getVendorBasedUUID() -> String? {
        #if os(iOS) || os(tvOS)
        // identifierForVendor is stable as long as at least one app from the vendor is installed
        let vendorId = UIDevice.current.identifierForVendor?.uuidString
        if let vendorId = vendorId {
            logger.debug("Retrieved vendor identifier: \(vendorId)")
            return vendorId
        }
        #endif

        #if os(macOS)
        // On macOS, we can use system hardware UUID (requires elevated permissions)
        // For now, we'll skip this as it requires special entitlements
        #endif

        return nil
    }

    /// Generate and store a completely new UUID
    private static func generateAndStoreNewUUID() -> String {
        let newUUID = UUID().uuidString
        try? keychainManager.storeDeviceUUID(newUUID)
        return newUUID
    }

    /// Create SHA256 hash of a string
    private static func sha256(_ string: String) -> String {
        guard let data = string.data(using: .utf8) else { return "" }

        var hash = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        data.withUnsafeBytes {
            _ = CC_SHA256($0.baseAddress, CC_LONG(data.count), &hash)
        }

        return hash.map { String(format: "%02x", $0) }.joined()
    }
}

// MARK: - Import CommonCrypto (for SHA256)
#if os(iOS) || os(tvOS) || os(watchOS)
import CommonCrypto
#elseif os(macOS)
import CommonCrypto
#endif

File: sdk/runanywhere-kotlin/src/commonMain/kotlin/com/runanywhere/sdk/foundation/PersistentDeviceIdentity.kt
package com.runanywhere.sdk.foundation

import com.runanywhere.sdk.storage.SecureStorage
import com.runanywhere.sdk.storage.createSecureStorage
import kotlinx.coroutines.runBlocking
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

/**
 * Comprehensive device identity manager that provides persistent UUIDs across app reinstalls
 * Uses multiple fallback mechanisms to ensure device identification persistence
 * Head-to-head translation from iOS PersistentDeviceIdentity
 */
object PersistentDeviceIdentity {

    private val logger = SDKLogger("PersistentDeviceIdentity")
    private val secureStorage: SecureStorage by lazy { createSecureStorage() }

    private const val DEVICE_UUID_KEY = "com.runanywhere.sdk.device.uuid"
    private const val DEVICE_FINGERPRINT_KEY = "com.runanywhere.sdk.device.fingerprint"

    @Volatile
    private var cachedDeviceUUID: String? = null

    /**
     * Get a persistent device UUID that survives app reinstalls
     * Uses a multi-layered approach for maximum persistence
     * Matches iOS PersistentDeviceIdentity.getPersistentDeviceUUID()
     */
    @OptIn(ExperimentalUuidApi::class)
    fun getPersistentDeviceUUID(): String {
        logger.debug("Attempting to retrieve persistent device UUID")

        // Return cached UUID if available
        cachedDeviceUUID?.let {
            logger.debug("Using cached device UUID")
            return it
        }

        return runBlocking {
            // Strategy 1: Try to get from persistent secure storage (survives app reinstalls)
            val storedUUID = secureStorage.getSecureString(DEVICE_UUID_KEY)
            if (storedUUID != null) {
                logger.info("Retrieved device UUID from persistent secure storage")
                cachedDeviceUUID = storedUUID
                return@runBlocking storedUUID
            }

            // Strategy 2: Use platform-specific vendor UUID if available
            val vendorUUID = getVendorBasedUUID()
            if (vendorUUID != null) {
                logger.info("Retrieved device UUID from vendor identifier")
                // Store this in persistent storage for future use
                try {
                    secureStorage.setSecureString(DEVICE_UUID_KEY, vendorUUID)
                } catch (e: Exception) {
                    logger.error("Failed to store vendor UUID", e)
                }
                cachedDeviceUUID = vendorUUID
                return@runBlocking vendorUUID
            }

            // Strategy 3: Generate new UUID and store persistently
            val newUUID = generateAndStoreNewUUID()
            logger.info("Generated new device UUID")
            return@runBlocking newUUID
        }
    }

    /**
     * Get device fingerprint for additional validation
     * This provides device characteristics that can help verify identity
     * Matches iOS PersistentDeviceIdentity.getDeviceFingerprint()
     */
    fun getDeviceFingerprint(): String {
        val components = mutableListOf<String>()

        try {
            // Get platform-specific device characteristics
            val deviceInfo = getPlatformDeviceInfo()

            // Hardware characteristics (these don't change)
            components.add("mem:${deviceInfo.totalMemory}")
            components.add("arch:${deviceInfo.architecture}")
            components.add("cores:${deviceInfo.coreCount}")
            components.add("model:${deviceInfo.deviceModel}")

            // OS info (major version only, as minor versions change)
            components.add("os:${deviceInfo.osMajorVersion}")

            // Create fingerprint hash
            val fingerprintString = components.joinToString("|")
            val fingerprint = sha256(fingerprintString)

            // Store fingerprint for future validation
            runBlocking {
                try {
                    secureStorage.setSecureString(DEVICE_FINGERPRINT_KEY, fingerprint)
                } catch (e: Exception) {
                    logger.error("Failed to store device fingerprint", e)
                }
            }

            return fingerprint
        } catch (e: Exception) {
            logger.error("Failed to generate device fingerprint", e)
            return "unknown"
        }
    }

    /**
     * Validate if a device UUID still represents the same physical device
     * Matches iOS PersistentDeviceIdentity.validateDeviceUUID()
     */
    fun validateDeviceUUID(uuid: String): Boolean {
        // For now, we trust any UUID that's properly formatted
        // In the future, this could compare against device fingerprints
        return uuid.length == 36 && uuid.contains("-")
    }

    /**
     * Clear device identity (for testing or privacy purposes)
     */
    suspend fun clearDeviceIdentity() {
        cachedDeviceUUID = null
        secureStorage.removeSecure(DEVICE_UUID_KEY)
        secureStorage.removeSecure(DEVICE_FINGERPRINT_KEY)
    }

    /**
     * Check if device identity exists
     */
    suspend fun hasDeviceIdentity(): Boolean {
        return secureStorage.containsSecure(DEVICE_UUID_KEY)
    }

    // MARK: - Private Methods

    /**
     * Get vendor-based UUID (platform-specific implementation)
     */
    private suspend fun getVendorBasedUUID(): String? {
        return try {
            getPlatformVendorUUID()
        } catch (e: Exception) {
            logger.debug("Failed to get vendor UUID: ${e.message}")
            null
        }
    }

    /**
     * Generate and store a completely new UUID
     */
    @OptIn(ExperimentalUuidApi::class)
    private suspend fun generateAndStoreNewUUID(): String {
        val newUUID = Uuid.random().toString()
        try {
            secureStorage.setSecureString(DEVICE_UUID_KEY, newUUID)
        } catch (e: Exception) {
            logger.error("Failed to store new UUID", e)
        }
        cachedDeviceUUID = newUUID
        return newUUID
    }

    /**
     * Create SHA256 hash of a string
     */
    private fun sha256(string: String): String {
        return try {
            // Platform-specific SHA256 implementation
            platformSha256(string)
        } catch (e: Exception) {
            logger.error("Failed to generate SHA256", e)
            string.hashCode().toString()
        }
    }
}

/**
 * Platform-specific device information for fingerprinting
 */
data class PlatformDeviceInfo(
    val totalMemory: Long,
    val architecture: String,
    val coreCount: Int,
    val deviceModel: String,
    val osMajorVersion: String
)

/**
 * Platform-specific functions - implemented in each platform module
 */
expect suspend fun getPlatformVendorUUID(): String?
expect fun getPlatformDeviceInfo(): PlatformDeviceInfo
expect fun platformSha256(input: String): String
